### ISSUE TITLE
@app.after_request not running when send_directory 304's

### ISSUE THREAD

**Issue by kafonek** [pallets/flask#3352]
### Expected Behavior
I am adding headers to the response of every request in my app (a [pep 503](https://www.python.org/dev/peps/pep-0503/) style web-app to serve out internally hosted python libraries).  The extra header shows up in almost every response.

```python
app = Flask(__name__)

@app.after_request
def extra_header(response):
    response.headers['foo'] = 'bar'
    return response

### pep 503 endpoint to serve pip installs
@app.route('/packages/<package_name>/<filename>')
def get_package(package_name, filename):
    ### logic to get library from internal databases
    ### directory, filename get defined here
    return send_from_directory(directory, filename)
```

Then the users run a `pip install index-url=<flask-app url>`.  Every request/response must include the custom header though or it is blocked by internal gates.

### Actual Behavior
When a user runs a `pip install 'internal_library>1.2'` or similar syntax without a `--no-cache-dir` option, pip sends a GET to the server with `If-None-Match` and `If-Modified-Since` headers.  When those match, the web-server responds with status code **304** and headers that include `ETag` and `Cache-Control`, but not the custom headers specified in the `after_request` function.

### Environment

* Python version: 3.5.1
* Flask version: 0.12.2
* Werkzeug version: 0.12.2

**Comment by davidism** [comment]
I can't reproduce this issue as described. Nothing about `send_file` would be able to bypass `after_request`. I ran a sample application and observed the custom header with the 304 status code in the response.

```python
from flask import Flask
from flask import send_from_directory

app = Flask(__name__)


@app.after_request
def special_header(response):
    response.headers['foo'] = 'bar'
    return response


@app.route("/")
def index():
    return send_from_directory(".", "LICENSE.rst")


if __name__ == '__main__':
    import requests

    rv = requests.get(
        "http://127.0.0.1:5000/static/LICENSE.rst",
        headers={"If-None-Match": '"1561391683.603-1475-2062749753"'},
    )
    print(rv.status_code)
    print(rv.headers['foo'])
```

```
$ FLASK_APP=example flask run
```

```
$ python example.py
304
bar
```

**Comment by davidism** [comment]
Also note that Flask 0.12 and Werkzeug 0.12 are no longer directly supported. The latest versions are Flask 1.1 and Werkzeug 0.15.

**Comment by kafonek** [comment]
@davidism thanks for the quick response and tips on upgrading versions, looks like the default loads on our plaform-as-a-service tooling are out of date.  I upgraded to Flask 1.1 and Wekzeug 0.15 and still see the odd behavior of missing the custom headers only when `send_from_directory` returns the 304.  However, when I create a toy example like your code has and run it on my local machine, I can't replicate the problem.

For posterity, the production app uses a handful of flask extensions including `flask_cors`, `flask_bootstrap`, and `flask_nav` but I couldn't trigger this weird behavior when I included them in the toy example either.

I'm happy to leave this closed for now since I can't replicate the problem outside of our platform-as-a-service environment and just including `add_etags=False` is a workaround solution.  If I can ever replicate the problem I'll re-open this issue.

Thanks again for the help.
